File: Emulator/src/main/content/region/asgarnia/port_sarim/quest/hunt/PiratesTreasure.java
package content.region.asgarnia.port_sarim.quest.hunt;

import content.region.asgarnia.port_sarim.quest.hunt.plugin.PiratesTreasurePlugin;
import core.game.component.Component;
import core.game.node.entity.player.Player;
import core.game.node.entity.player.link.quest.Quest;
import core.game.node.item.GroundItemManager;
import core.game.node.item.Item;
import core.plugin.ClassScanner;
import core.plugin.Initializable;
import org.rs.consts.Components;
import org.rs.consts.Items;
import org.rs.consts.Quests;
import org.rs.consts.Vars;

import static core.api.ContentAPIKt.removeAttribute;

/**
 * The type Pirates treasure.
 */
@Initializable
public final class PiratesTreasure extends Quest {

    /**
     * The constant MESSAGE_COMPONENT.
     */
    public static final Component MESSAGE_COMPONENT = new Component(Components.BLANK_SCROLL_222);

    /**
     * The constant CASKET_REWARDS.
     */
    public static final Item[] CASKET_REWARDS = new Item[]{new Item(Items.COINS_995, 450), new Item(Items.GOLD_RING_1635), new Item(Items.EMERALD_1605)};

    /**
     * The constant KARAMJAN_RUM.
     */
    public static final Item KARAMJAN_RUM = new Item(Items.KARAMJAN_RUM_431);

    /**
     * The constant KEY.
     */
    public static final Item KEY = new Item(Items.CHEST_KEY_432);

    /**
     * The constant PIRATE_MESSAGE.
     */
    public static final Item PIRATE_MESSAGE = new Item(Items.PIRATE_MESSAGE_433);

    /**
     * The constant CASKET.
     */
    public static final Item CASKET = new Item(Items.CASKET_7956);

    /**
     * Instantiates a new Pirates treasure.
     */
    public PiratesTreasure() {
        super(Quests.PIRATES_TREASURE, 23, 22, 2, Vars.VARP_QUEST_PIRATES_TREASURE_PROGRESS_71, 0, 1, 4);
    }

    @Override
    public Quest newInstance(Object object) {
        ClassScanner.definePlugin(new PiratesTreasurePlugin());
        return this;
    }

    @Override
    public void drawJournal(Player player, int stage) {
        super.drawJournal(player, stage);
        switch (stage) {
            case 0:
                player.getPacketDispatch().sendString(BLUE + "I can start this quest by speaking to " + RED + "Redbeard Frank " + BLUE + "who", 275, 4 + 7);
                player.getPacketDispatch().sendString(BLUE + "is at " + RED + " Port Sarim.", 275, 5 + 7);
                player.getPacketDispatch().sendString(BLUE + "There aren't any requirements for this quest.", 275, 7 + 7);
                break;
            case 10:
                if (!player.getInventory().containsItem(KARAMJAN_RUM)) {
                    line(player, BLUE + "I have spoken to " + RED + " Redbeard Frank." + BLUE + " He has agreed to tell me", 4 + 7);
                    line(player, BLUE + "the location of some " + RED + " treasure " + BLUE + "for some " + RED + "Karamjan rum.", 5 + 7);
                    line(player, BLUE + "I have taken employment on the " + RED + "banana plantation " + BLUE + ", as the", 6 + 7);
                    line(player, RED + "Customs Officers " + BLUE + "might not notice the " + RED + "rum " + BLUE + "if it is covered", 7 + 7);
                    line(player, BLUE + "in " + RED + "bananas.", 8 + 7);
                    line(player, BLUE + "Now all I need is some " + RED + " rum " + BLUE + "to hide in the next crate", 10 + 7);
                    line(player, BLUE + "destined for " + RED + "Wydin's store" + BLUE + "...", 11 + 7);
                } else {
                    line(player, BLUE + "I have spoken to " + RED + "Redbeard Frank " + BLUE + ". He has agreed to tell me", 5 + 7);
                    line(player, BLUE + "the location of some " + RED + "treasure " + BLUE + "for some " + RED + "Karamjan rum.", 6 + 7);
                    line(player, BLUE + "I have the " + RED + "Karamjan rum. " + BLUE + "I should take it to " + RED + "Redbeard Frank.", 8 + 7);
                }
                break;
            case 20:
                line(player, "<str>I have smuggled some rum off Karamja, and retrieved it", 4 + 7);
                line(player, "<str>from the back room of Wydin's shop.", 5 + 7);
                line(player, BLUE + "I have given the rum to " + RED + "Redbeard Frank." + BLUE + " He has told me", 7 + 7);
                line(player, BLUE + "that the " + RED + "treasure " + BLUE + "is hidden in the chest in the upstairs", 8 + 7);
                line(player, BLUE + "room of the " + RED + "Blue Moon Inn " + BLUE + "in " + RED + "Varrock.", 9 + 7);
                if (player.getAttributes().containsKey("pirate-read")) {
                    line(player, BLUE + "The note reads " + RED + "'Visit the city of the White Knights. In the", 11 + 7);
                    line(player, RED + "park, Saradomin points to the X which marks the spot.'", 12 + 7);
                    if (!player.getInventory().containsItem(PIRATE_MESSAGE) && !player.getBank().containsItem(PIRATE_MESSAGE)) {
                        line(player, BLUE + "It's a good job I remembered that, as I have lost the " + RED + "note.", 14 + 7);
                    }
                    return;
                }
                if (player.getInventory().containsItem(PIRATE_MESSAGE)) {
                    line(player, BLUE + "I have opened the chest in the " + RED + "Blue Moon" + BLUE + ", and found a", 11 + 7);
                    line(player, RED + "note " + BLUE + "inside. I think it will tell me where to dig.", 12 + 7);
                    return;
                }
                if (player.getInventory().containsItem(KEY)) {
                    line(player, BLUE + "I have a " + RED + "key " + BLUE + "that can be used to unlock the chest that", 11 + 7);
                    line(player, BLUE + "holds the treasure.", 12 + 7);
                } else {
                    line(player, BLUE + "I have lost the " + RED + "key " + BLUE + "that " + RED + "Redbeard Frank " + BLUE + "gave me. I should", 11 + 7);
                    line(player, BLUE + "see if he has another.", 12 + 7);
                }
                break;
            case 100:
                line(player, "<str>The note reads 'Visit the city of the White Knights. In the", 4 + 7);
                line(player, "<str>park, Saradomin points to the X which marks the spot.'", 5 + 7);
                line(player, "<col=FF0000>QUEST COMPLETE!", 7 + 7);
                line(player, BLUE + "I've found a treasure, gained 2 Quest Points and gained", 9 + 7);
                line(player, BLUE + "access to the Pay-fare option to travel to and from", 10 + 7);
                line(player, BLUE + "Karamja!", 11 + 7);
                break;
        }
    }

    @Override
    public void finish(Player player) {
        super.finish(player);
        player.getPacketDispatch().sendString("2 Quest Points", Components.QUEST_COMPLETE_SCROLL_277, 8 + 2);
        player.getPacketDispatch().sendString("One-Eyed Hector's Treasure", Components.QUEST_COMPLETE_SCROLL_277, 9 + 2);
        player.getPacketDispatch().sendString("Chest", Components.QUEST_COMPLETE_SCROLL_277, 10 + 2);
        player.getPacketDispatch().sendString("You can also use the Pay-", Components.QUEST_COMPLETE_SCROLL_277, 11 + 2);
        player.getPacketDispatch().sendString("fare option to go to and from", Components.QUEST_COMPLETE_SCROLL_277, 12 + 2);
        player.getPacketDispatch().sendString("Karamja", Components.QUEST_COMPLETE_SCROLL_277, 13 + 2);
        player.getPacketDispatch().sendString("You have completed the Pirate's Treasure Quest!", Components.QUEST_COMPLETE_SCROLL_277, 2 + 2);
        player.getPacketDispatch().sendItemZoomOnInterface(Items.CASKET_7956, 230, Components.QUEST_COMPLETE_SCROLL_277, 3 + 2);
        removeAttribute(player, "gardener-attack");
        removeAttribute(player, "pirate-read");
        if (!player.getInventory().add(CASKET)) {
            GroundItemManager.create(CASKET, player);
        }
    }

}

File: Emulator/src/main/content/region/asgarnia/portsarim/quest/hunt/PiratesTreasure.kt
package content.region.asgarnia.portsarim.quest.hunt

import core.api.*
import core.game.node.entity.player.Player
import core.game.node.entity.player.link.quest.Quest
import core.game.node.item.Item
import core.plugin.Initializable
import shared.consts.Components
import shared.consts.Items
import shared.consts.Quests
import shared.consts.Vars

@Initializable
class PiratesTreasure : Quest(
    Quests.PIRATES_TREASURE,
    23, 22, 2,
    Vars.VARP_QUEST_PIRATES_TREASURE_PROGRESS_71,
    0, 1, 4
) {

    companion object {
        private const val KARAMJAN_RUM = Items.KARAMJAN_RUM_431
        private const val KEY = Items.CHEST_KEY_432
        private const val PIRATE_MESSAGE = Items.PIRATE_MESSAGE_433
        private const val CASKET = Items.CASKET_7956

        val CASKET_REWARDS: Array<Item> = arrayOf(
            Item(Items.COINS_995, 450),
            Item(Items.GOLD_RING_1635, 1),
            Item(Items.EMERALD_1605, 1)
        )
    }

    override fun newInstance(`object`: Any?): Quest {
        return this
    }

    override fun drawJournal(player: Player, stage: Int) {
        super.drawJournal(player, stage)
        var line = 11

        when (stage) {
            0 -> {
                line(player, "I can start this quest by speaking to !!Redbeard Frank?? who", line++)
                line(player, "is at !!Port Sarim??.", line++)
                line(player, "There aren't any requirements for this quest.", line)
            }

            10 -> {
                if (!inInventory(player, KARAMJAN_RUM)) {
                    line(player, "I have spoken to !!Redbeard Frank??. He has agreed to tell me", line++)
                    line(player, "the location of some !!treasure?? for some !!Karamjan rum??.", line++)
                    line(player, "I have taken employment on the !!banana plantation??,", line++)
                    line(player, "as the !!Customs Officers?? might not notice the rum if", line++)
                    line(player, "it is covered in !!bananas??.", line++)
                    line++
                    line(player, "Now all I need is some !!rum?? to hide in the next crate", line++)
                    line(player, "destined for !!Wydin's store??...", line)
                } else {
                    line(player, "I have spoken to !!Redbeard Frank??. He has agreed to tell me", line++)
                    line(player, "the location of some !!treasure?? for some !!Karamjan rum??.", line++)
                    line++
                    line(player, "I have the !!Karamjan rum??. I should take it to !!Redbeard Frank??.", line)
                }
            }

            20 -> {
                line(player, "<str>I have smuggled some rum off Karamja, and retrieved it", line++)
                line(player, "<str>from the back room of Wydin's shop.", line++)
                line++
                line(player, "I have given the rum to !!Redbeard Frank??. He has told me", line++)
                line(player, "that the !!treasure?? is hidden in the chest in the upstairs", line++)
                line(player, "room of the !!Blue Moon Inn?? in !!Varrock??.", line++)

                when {
                    player.attributes.containsKey("pirate-read") -> {
                        line++
                        line(player, "The note reads !!'Visit the city of the White Knights. In the", line++)
                        line(player, "park, Saradomin points to the X which marks the spot.'??", line++)
                        if (!inInventory(player, PIRATE_MESSAGE) && !inBank(player, PIRATE_MESSAGE)) {
                            line++
                            line(player, "It's a good job I remembered that, as I have lost the !!note??.", line)
                        }
                    }

                    inInventory(player, PIRATE_MESSAGE) -> {
                        line++
                        line(player, "I have opened the chest in the !!Blue Moon??, and found a", line++)
                        line(player, "!!note?? inside. I think it will tell me where to dig.", line)
                    }

                    inInventory(player, KEY) -> {
                        line++
                        line(player, "I have a !!key?? that can be used to unlock the chest that", line++)
                        line(player, "holds the treasure.", line)
                    }

                    else -> {
                        line++
                        line(player, "I have lost the !!key?? that !!Redbeard Frank?? gave me. I should", line++)
                        line(player, "see if he has another.", line)
                    }
                }
            }

            100 -> {
                line(player, "<str>The note reads 'Visit the city of the White Knights. In the", line++)
                line(player, "<str>park, Saradomin points to the X which marks the spot.'", line++)
                line++
                line(player, "%%QUEST COMPLETE!&&", line++)
                line++
                line(player, "I've found a treasure, gained 2 Quest Points and gained", line++)
                line(player, "access to the !!Pay-fare?? option to travel to and from", line++)
                line(player, "!!Karamja??!", line)
            }
        }
    }

    override fun finish(player: Player) {
        super.finish(player)

        var line = 10

        sendItemZoomOnInterface(player, Components.QUEST_COMPLETE_SCROLL_277, 5, Items.CASKET_7956)
        drawReward(player, "2 Quest Points", line++)
        drawReward(player, "One-Eyed Hector's Treasure", line++)
        drawReward(player, "Chest", line++)
        drawReward(player, "You can also use the Pay-", line++)
        drawReward(player, "fare option to go to and from", line++)
        drawReward(player, "Karamja", line)

        removeAttribute(player, "gardener-attack")
        removeAttribute(player, "pirate-read")

        addItemOrDrop(player, CASKET)
        updateQuestTab(player)
    }
}
